### PR TITLE
Rewriting plugin postprocessor

### PR DESCRIPTION
**Pull request by jessicafalk**
**Contributions**: We are not currently taking public contributions - see our [contributions](https://docs.improbable.io/unity/alpha/contributing) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
trying to make it simpler and smaller other drive-by fixes. 
Before the Plugin importer was very generic, however we only want to use this for our own internal Improbable plugins at the moment. We know the structure of these exactly, therefore we can make everything a lot simpler. 

The basic structure for Plugins:
- Core plugin (ie worker library)?  We need a separate own for every platform and cpu architecture
- SDK plugin (ie. worker sdk)? As it's a DLL we can reuse it for all platforms except for iOS. iOS needs a statically linked one.

If we ever need to make it more extensible or more generic, we can revisit the plugin importer and would probably need to move it into it's own feature module.

#### Tests
ran it locally.
ran a cloud deployment, everything spawned correctly

#### Documentation
How is this documented (for example: release note, upgrade guide, feature page, in-code documentation)?

#### Primary reviewers
If your change will take a long time to review, you can name at most two primary reviewers who are ultimately responsible for reviewing this request. @ mention them.

### CODE REVIEW

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        ///     Sets plugin platform compatibility based on directory structure
+        /// </summary>
+        internal static void SetPluginsCompatibility()
+        {
+            AssetDatabase.StartAssetEditing();
+            foreach (var pluginDirectoryCompatibility in PluginsCompatibilityList)
+            {
+                if (!Directory.Exists(pluginDirectoryCompatibility.PluginPath))
+                {
+                    Debug.Log("directory doesn't exist " + pluginDirectoryCompatibility.PluginPath);
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs]: Nit: "**D**irectory"

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+                    continue;
+                }
+
+                var pluginPaths = AssetDatabase.FindAssets(string.Empty, new[] { pluginDirectoryCompatibility.PluginPath });
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs]: driveby opportunity: `pluginPaths` is incorrect, it should be `pluginGuids`

Or, `pluginGuidsInFolder`?

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+                foreach (var pluginPath in pluginPaths)
+                {
+                    var plugin = AssetImporter.GetAtPath(AssetDatabase.GUIDToAssetPath(pluginPath)) as PluginImporter;
+                    var needsReimport = false;
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs]: driveby: can move this below the continue check

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginDirectoryCompatibility.cs [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+        ARMv7,
+        Arm64,
+        X86_64,
+        X86,
+        Unused,
+    }
+
+    internal class PluginDirectoryCompatibility
+    {
+        public readonly bool AnyPlatformCompatible;
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginDirectoryCompatibility.cs]: How about `CompatibleWithAnyPlatform`?

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginDirectoryCompatibility.cs [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+        public readonly bool IsEditorCompatible;
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginDirectoryCompatibility.cs]: How about `CompatibleWithEditor`?

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginDirectoryCompatibility.cs (hunk 4 above)
Comment (firtina-improbable) [reply workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginDirectoryCompatibility.cs]: for consistency

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs [hunk 6]
```diff
@@ -0,0 +1,135 @@
...
+                        needsReimport = true;
+                    }
+
+                    if (pluginDirectoryCompatibility.IsEditorCompatible != plugin.GetCompatibleWithEditor())
+                    {
+                        plugin.SetCompatibleWithEditor(pluginDirectoryCompatibility.IsEditorCompatible);
+                        needsReimport = true;
+                    }
+
+                    if (pluginDirectoryCompatibility.CompatiblePlatform > 0 && !plugin.GetCompatibleWithPlatform(pluginDirectoryCompatibility.CompatiblePlatform))
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs]: != 0 may be more appropriate

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginDirectoryCompatibility.cs [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+                { CPUType.X86, "x86" },
+                { CPUType.X86_64, "x86_64" },
+            };
+
+        public PluginDirectoryCompatibility(
+            PluginType pluginType,
+            BuildTarget buildTarget,
+            CPUType cpuType = CPUType.Unused,
+            bool isEditorCompatible = true
+            )
```
Comment (firtina-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginDirectoryCompatibility.cs]: nit: mismatching indent for `)`, also should the opening one be in a new line, or should this one be next to `true`?

File: workers/unity/Assets/Playground/Scripts/UI/ConnectionScreenController.cs [hunk 8]
```diff
@@ -96,7 +98,7 @@ public void OnConnectionSucceeded()
         public void OnConnectionFailed(string connectionError)
         {
             UnityObjectDestroyer.Destroy(worker);
-            errorMessage = $"Connection failed. Please check the IP address entered.\n{connectionError}";
+            errorMessage = $"Connection failed. Please check the IP address entered.\nError received by SpatialOS:\n{connectionError}";
```
Comment (jared-improbable) [inline workers/unity/Assets/Playground/Scripts/UI/ConnectionScreenController.cs]: Perhaps "SpatialOS error message:" ?

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginCompatibilitySetting.cs [hunk 9]
```diff
@@ -0,0 +1,87 @@
...
+                { CPUType.X86_64, "x86_64" },
+            };
+
+        public PluginCompatibilitySetting(
+            PluginType pluginType,
+            BuildTarget compatiblePlatform,
+            CPUType cpuType = CPUType.Unused,
+            bool compatibleWithEditor = true)
+        {
+
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginCompatibilitySetting.cs]: Nit: blank line

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginCompatibilitySetting.cs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+            bool compatibleWithEditor = true)
+        {
+            
+            PluginType = pluginType;
+            PluginPath = Path.Combine(DownloadCoreSdk.ImprobablePluginsPath, pluginType.ToString(), BuildTargetToFolder[compatiblePlatform]);
+            CompatibleWithAnyPlatform = false;
+            CompatibleWithEditor = compatibleWithEditor;
+            CompatiblePlatform = compatiblePlatform;
+            CPU = string.Empty;
+            if (CPUType.Unused != cpuType)
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginCompatibilitySetting.cs]: Maybe it's Friday morning, but I had trouble reading this reversed comparison

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs [hunk 11]
```diff
@@ -0,0 +1,133 @@
...
+                BuildTarget.Android,
+                CPUType.X86
+                ),
+            new PluginCompatibilitySetting(
+                PluginType.Core,
+                BuildTarget.iOS
+                ),
+            new PluginCompatibilitySetting(
+                PluginType.Sdk,
+                new List<BuildTarget> { BuildTarget.iOS }
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs]: Any reason this is a `List`?

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+                ),
+            new PluginCompatibilitySetting(
+                PluginType.Sdk,
+                BuildTarget.iOS,
+                CPUType.Unused,
+                false
```
Comment (jared-improbable) [inline workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs]: What's `false`?

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs (hunk 11 above)
Comment (jessicafalk) [reply workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs]: just in case we need to add more platforms in the feature and to differentiate between compatible and incompatible platforms.
could also be an array or I use static methods to initalize it. What would you prefer?

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs (hunk 12 above)
Comment (jessicafalk) [reply workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs]: added name identifier

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginCompatibilitySetting.cs (hunk 10 above)
Comment (jessicafalk) [reply workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginCompatibilitySetting.cs]: rewrote it a bit ;)

File: workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs (hunk 11 above)
Comment (jared-improbable) [reply workers/unity/Packages/com.improbable.gdk.tools/Plugins/PluginPostprocessor.cs]: This is fine, I was just curious because it stood out from the other entries.
